text/0323-array-functions.md

```diff
@@ -0,0 +1,130 @@
+
+All of these problems go away when using array functions instead of array methods.
+
+It's also worth noting that when you using fastboot you are required to have `Array` prototype extensions disabled currently and thus forced to deal with these issues.
+
+## Detailed design
+
+For each method in the `MutableArray` mixin we will export a similarly named function from `@ember/array`.
+
+For example, the method `array.objectAt(index)` will map to the function `objectAt(array, index)` with implementation
+
+```js
+function objectAt(array, index) {
+  if (Array.isArray(array)) {
+    return array[index];
+  } else {
+    return array.objectAt(index);
+  }
+}
+```
+
+As you can see, the native array case is handled explicitly in order to avoid assuming that the global `Array` prototype has been extended, but otherwise we just delegate to the `objectAt` method.
+
+For reference, below is an expandable list of all array functions we will need.
+
+<details>
+  <summary>List of array functions (49)</summary>
+  <code>
+    <ul>
+      <li>addArrayObserver</li>
+      <li>addObject</li>
+      <li>addObjects</li>
+      <li>any</li>
+      <li>arrayContentDidChange</li>
+      <li>arrayContentWillChange</li>
+      <li>clear</li>
+      <li>compact</li>
+      <li>every</li>
+      <li>filter</li>
+      <li>filterBy</li>
+      <li>find</li>
+      <li>findBy</li>
+      <li>forEach</li>
+      <li>getEach</li>
+      <li>includes</li>
+      <li>indexOf</li>
+      <li>insertAt</li>
+      <li>invoke</li>
+      <li>isAny</li>
+      <li>isEvery</li>
+      <li>lastIndexOf</li>
+      <li>map</li>
+      <li>mapBy</li>
+      <li>objectAt</li>
+      <li>objectsAt</li>
+      <li>popObject</li>
+      <li>pushObject</li>
+      <li>pushObjects</li>
+      <li>reduce</li>
+      <li>reject</li>
+      <li>rejectBy</li>
+      <li>removeArrayObserver</li>
+      <li>removeAt</li>
+      <li>removeObject</li>
+      <li>removeObjects</li>
+      <li>replace</li>
+      <li>reverseObjects</li>
+      <li>setEach</li>
+      <li>setObjects</li>
+      <li>shiftObject</li>
+      <li>slice</li>
+      <li>sortBy</li>
+      <li>toArray</li>
+      <li>uniq</li>
+      <li>uniqBy</li>
+      <li>unshiftObject</li>
+      <li>unshiftObjects</li>
+      <li>without</li>
+    </ul>
+  </code>
+</details>
+
+## How we teach this
+
+This functions must obviously be documented under the `@ember/array` package in the API docs.
+
+We should encourage people to use the phrase "array functions" to describe these functions and "array methods" to describe the method versions.
+
+We should encourage their usage in addons and fastboot apps instead of using `Ember.A`.
+
+In the future, we may want to encourage their usage in all apps so that we can deprecate extending the global `Array` prototype, but that is out of scope for this RFC.
+
+## Drawbacks
```

Comment:
I think a drawback here is that these will look like "common array functions". Folks might not understand why they need to use the ember version and try to swap it for something like lodash.

Does that mean that this will work...

```js
import { map } from 'ember-array-functions';

map(arr, e => e * 2);
```

But this would not?

```js
import { map } from 'lodash';

map(arr, e => e * 2);
```

That might be confusing if we start pushing application developers down this road.

Reply:
see https://github.com/emberjs/rfcs/pull/323#discussion_r176941217

"Array" in Ember is essentially an interface, that can have multiple implementations. One implementation is native arrays where lodash will work fine. For other implementations only the Ember functions will work since they are aware of that interface.